qutip/tests/test_interpolate.py

```diff
@@ -30,325 +30,123 @@
+        (qutip.mesolve, 'mesolve'),
+        (functools.partial(qutip.mcsolve, ntraj=500), "mcsolve"),
+        (qutip.brmesolve, 'brmesolve')
+    ]
+    coefficients = [
+        ([_case_real_spline], "real-spline"),
+        ([_case_real_spline, _case_real_string], "real-spline,string"),
+        ([_case_real_spline, _case_real_function], "real-spline,function"),
+        ([_case_real_function], "real-function"),
+        ([_case_complex_spline, _case_complex_string],
+            "complex-spline,string"),
+    ]
+
+    def _valid_case(solver_id, coefficient_id):
+        invalids = [
+            "brmesolve" in solver_id and "function" in coefficient_id,
+        ]
+        return not any(invalids)
+
+    def _make_case(solver, solver_id, coefficient, coefficient_id):
+        marks = []
+        if "brmesolve" in solver_id:
+            marks.append(pytest.mark.requires_cython)
+        id_ = "-".join([solver_id, coefficient_id])
+        return pytest.param(solver, coefficient, id=id_, marks=marks)
+
+    cases = [_make_case(solver, solver_id, coefficient, coefficient_id)
+             for solver, solver_id in solvers
+             for coefficient, coefficient_id in coefficients
+             if _valid_case(solver_id, coefficient_id)]
+    metafunc.parametrize(["solver", "coefficients"], cases)
 
 
 @pytest.mark.slow
-@unittest.skipIf(not Cython_OK, 'Cython not found or version too low.')
-def test_interpolate_brevolve3():
-    """
-    Interpolate: brmesolve c_op as interp (str)
-    """
-    N = 10  # number of basis states to consider
-    a = destroy(N)
-    H = a.dag() * a
-    psi0 = basis(N, 9)  # initial state
-    kappa = 0.2  # coupling to oscillator
-    tlist = np.linspace(0, 10, 100)
-
-    f = lambda t: np.sqrt(kappa*np.exp(-t))
-
-    S = Cubic_Spline(tlist[0],tlist[-1],f(tlist))
-    c_ops = [[a, S]]
-    brdata = brmesolve(H, psi0, tlist, a_ops=[], e_ops=[a.dag() * a], c_ops=c_ops)
-    expt = brdata.expect[0]
-    actual_answer = 9.0 * np.exp(-kappa * (1.0 - np.exp(-tlist)))
-    avg_diff = np.mean(abs(actual_answer - expt) / actual_answer)
-    assert_(avg_diff < 1e-4)
-
-
-if __name__ == "__main__":
-    run_module_suite()
+def test_usage_in_solvers(solver, coefficients):
+    """
+    Test that the Cubic_Spline can be used as a time-dependent argument to all
+    three principle solvers, both alone and in combination with other forms of
+    time-dependence.
+    """
+    size = 10
+    a = qutip.destroy(size)
+    state = qutip.basis(size, 1)
+    times = np.linspace(0, 5, 50)
+    e_ops = [a.dag() * a]
+    H_expected = [a.dag()*a]
+    H_test = H_expected.copy()
+    operator = a**2 + a.dag()**2
+    for string, other, type_ in coefficients:
+        if type_ == 'spline':
+            other = qutip.Cubic_Spline(times[0], times[-1], other(times))
+        H_expected.append([operator, string])
+        H_test.append([operator, other])
+    expected = solver(H_expected, state, times, e_ops=e_ops).expect[0]
```

Comment:
`mesolve` and `mcsolve` needs `c_ops`, otherwise they just call `sesolve`.

Reply:
As far as I remember, I didn't add these tests, I just converted the pre-existing ones.  I noticed this behaviour, but I'd left it in these commits, because I wanted to be certain I wasn't accidentally removing any.

I'll convert them to tests of a type "`mcsolve` calls `sesolve` is `c_ops is None`" in another commit.

Reply:
Looking at it, the original tests also did not have any c_ops... The fallback to `sesolve` was probably added after these tests. For the test, could you add a `qutip.qzero` as c_ops. It would ensure that the right solver is used.